File: ruby/zipper/zipper.rb
class Zipper
  include Comparable

  def initialize(root_tree)
    @root_tree = root_tree
    @current_node = @root_tree
    @parents = []
  end

  def to_tree
    @root_tree
  end

  def left
    @parents << @current_node
    @current_node = @current_node.left
    @current_node.nil? ? nil : self
  end

  def right
    @parents << @current_node
    @current_node = @current_node.right
    @current_node.nil? ? nil : self
  end

  def up
    return nil if @parents.empty?
    @current_node = @parents.pop
    self
  end

  def value
    @current_node.value
  end

  def set_value(value)
    @current_node.value = value
    self
  end

  def set_left(left)
    @current_node.left = left
    self
  end

  def set_right(right)
    @current_node.right = right
    self
  end

  def <=>(other)
    to_tree <=> other.to_tree
  end

  def self.from_tree(nodes)
    self.new(nodes)
  end
end

class Node
  include Comparable

  attr_accessor :value, :left, :right

  def initialize(value, left, right)
    @value = value
    @left = left
    @right = right
  end

  def <=>(other)
    @value <=> other.value
  end
end

File: python/zipper/zipper.py
class Zipper(object):
    @staticmethod
    def from_tree(tree):
        return Zipper(tree)

    def __init__(self, tree):
        self.tree = tree
        self.tree_ptr = tree
        self.parents = []

    def value(self):
        return self.tree_ptr['value']

    def set_value(self, value):
        self.tree_ptr['value'] = value
        return self

    def left(self):
        self.parents.append(self.tree_ptr)
        if self.tree_ptr['left'] is None:
            return None
        else:
            self.tree_ptr = self.tree_ptr['left']
            return self

    def set_left(self, left):
        self.tree_ptr['left'] = left
        return self

    def right(self):
        self.parents.append(self.tree_ptr)
        if self.tree_ptr['right'] is None:
            return None
        else:
            self.tree_ptr = self.tree_ptr['right']
            return self

    def set_right(self, right):
        self.tree_ptr['right'] = right
        return self

    def up(self):
        if len(self.parents) == 0:
            return None
        self.tree_ptr = self.parents.pop()
        return self

    def to_tree(self):
        self.tree_ptr = self.tree
        self.parents = []
        return self.tree
